CONTRIBUTING.md

```diff
@@ -0,0 +1,316 @@
+9. [Add](https://git-scm.com/docs/git-add) the changes,
+   [commit](https://git-scm.com/docs/git-commit) them
+   and [push](https://git-scm.com/docs/git-push) the result, like this:
+
+        git add widgets/bar_baz.lua README.md
+        git commit -m '[bar_baz] Add widget type'
+        git add helpers.lua Changes.md
+        git commit -m '[helpers] Fix foo'
+        git push
+
+10. Finally, [create a pull request](https://help.github.com/articles/creating-a-pull-request).
+    We'll then review and merge it.
+
+In any case, thank you very much for your contributions!
+
+## Coding Conventions
+
+This section introduces a guideline for writing idiomatic, robust
+and future-proof widget type code.
+
+### Whitespces in Expressions and Statements
+
+Avoid extraneous whitespace in the following situations:
+
+* Immediately inside parentheses or brackets.  Braces, however, are exceptions
+  to this rule:
+
+  ```lua
+  foo(bar[1], { baz = 2 })      -- yes
+  foo( bar[ 1 ], {baz = 2} )    -- no
+  ```
+
+* Immediately before a comma, semicolon, or colon.
+* Immediately before the open parenthesis, braces, quote, etc.
+  that starts the argument list of a function call; or the open braket
+  that starts an indexing.  In other words, prefer these:
+
+  ```lua
+  foo(bar, baz)
+  foo{ bar, baz }
+  foo"bar"
+  foo[[bar]]
+  foo[bar]
+  ```
+
+* Trailing at the end of line or (newline) at the end of file.
+
+Always surround these binary operators with a single space on either side:
+assignment (`=`), comparisons, Booleans (`and`, `or`, `not`).
+If operators with different priorities are used, consider adding whitespace
+around the operators with the lowest priorities. Use your own judgment;
+however, never use more than one space, and always have
+the same amount of whitespace on both sides of a binary operator.
+
+### Indentation
+
+Use 4 *spaces* per indentation level.
+
+Continuation lines should align wrapped elements either vertically
+inside parentheses, brackets and braces, or using a hanging indent
+(the opening parenthesis of a parenthesized statement is the last
+non-whitespace character of the line, with subsequent lines being indented
+until the closing parenthesis), e.g.
+
+```lua
+-- Vertically aligned
+long_function_call{ foo, bar,
+                    baz }
+
+-- Hanging indentation
+long_function_call(
+    foo, bar
+    baz)
+```
+
+The closing brace or braket on multiline constructs may either line up under
+the first character of the line that starts the construct, as in:
+
+```lua
+long_function_call{
+    foo = 1, bar = 2,
+    baz = 3,
+}
+```
+
+In this case, and this case only, the trailing comma is acceptable
+to avoid diff noises when more values are added,
+but since Vicious often deal with system APIs which rarely ever change,
+it's occasionally helpful to do so.
+
+Trailing right parentheses, however, are not allowed.
+
+### Maximum Line Length
+
+Limit all *code* lines to a maximum of 80 characters.
```

Comment:
I think this can be relaxed a bit. The code base also contains examples exceeding the limit. This is what systemd for example proposes: https://github.com/systemd/systemd/blob/master/docs/CODING_STYLE.md#formatting

Reply:
I'm not sure about other people's screen, but my 1600x900 is just enough to fit two 80-col splits.  Maybe systemd's situation is different because their devs tends to use IDEs full-screen w/o splitting?  Of course I'll try to rephrase the guide to make it less harsh though.

BTW in this branch, only comments (which often contain long URLs) and long multi-line strings exceed the 80-col limit; but those are not handled by luacheck `max_code_line_length`.

Reply:
At some point long lines just become less readable and also github won't show long lines fully when reviewing pull requests.